### PR TITLE
Fix service.running when test=true

### LINKED ISSUE #34019: test=True reports failure when using file.append or similar functions
### Description of Issue/Question

Handling of file.append and required can be improved when using 'test=True' mode. When executing the state file with test=True, there are numerous errors compared to zero error for an actual execution.
### Setup

(Please provide relevant configs and/or SLS files (Be sure to remove sensitive info).)
Gist of a State file:

```
md5check_create:
  file.touch:
    - name: /home/{{ grains['id'] }}/jars.md5check

{% for key,jar in pillar['jars'].iteritems() %}
/home/{{ grains['id'] }}/{{ jar['directory'] }}/{{ jar['name'] }}-{{ jar['version'] }}.jar:
  cmd.run:
    - name: curl https://somestite/nexus/content/groups/prod/{{ jar['group']|replace('.', '/') }}/{{ jar['name'] }}/{{ jar['version'] }}/{{ jar['name'] }}-{{ jar['version'] }}.jar -H 'Authorization:Basic base64encoded' -o {{ jar['name'] }}-{{ jar['version'] }}.jar && echo {{ jar['md5'] }}
    - timeout: 30
    - cwd: /home/{{ grains['id'] }}/{{ jar['directory'] }}
    - runas: {{ grains['id'] }}
    - umask: 027

/home/{{ grains['id'] }}/{{ jar['directory'] }}/{{ jar['name'] }}-{{ jar['version'] }}.jar.md5:
  file.append:
    - name: /home/{{ grains['id'] }}/jars.md5check
    - require:
      - file: md5check_create
    - text: {{ jar['md5'] }}  /home/{{ grains['id'] }}/{{ jar['directory'] }}/{{ jar['name'] }}-{{ jar['version'] }}.jar

{% endfor %}

```
### Steps to Reproduce Issue

(Include debug logs if possible and relevant.)

**salt-ssh '_server_' --priv ~/.ssh/id_rsa state.sls somestate.withappend test=True**

Logs for test=True

```
----------
          ID: md5check_create
    Function: file.touch
        Name: /home/logged_user/jars.md5check
      Result: None
     Comment: File /home/logged_user/jars.md5check is set to be created
     Started: 21:37:57.324200
    Duration: 0.365 ms
     Changes:
----------
          ID: /home/logged_user/runtime/lib/artifact-1.0.10.1.jar
    Function: cmd.run
        Name: curl https://somesite/nexus/content/groups/prod/group/artifact/1.0.10.1/artifact-1.0.10.1.jar -H 'Authorization:Basic base64encoded' -o artifact-1.0.10.1.jar && echo md5checksumvalue
      Result: None
     Comment: Command "curl https://somesite/nexus/content/groups/prod/group/artifact/1.0.10.1/artifact-1.0.10.1.jar -H 'Authorization:Basic base64encoded' -o artifact-1.0.10.1.jar && echo md5checksumvalue" would have been executed
     Started: 21:37:57.326817
    Duration: 0.452 ms
     Changes:
----------
          ID: /home/logged_user/runtime/lib/artifact-1.0.10.1.jar.md5
    Function: file.append
        Name: /home/logged_user/jars.md5check
      Result: False
     Comment: /home/logged_user/jars.md5check: file not found
     Started: 21:37:57.331353
    Duration: 0.468 ms
     Changes:
----------

```

I would hope if we could give a bit more clear error message to user.  For example, the above comment could say: as executing in test mode, /home/logged_user/jars.md5check: file was not actually created.

And then in future versions, if we could mock the output of supported commands. Or atleast mention, command will not show accurate result when running in test mode.
### Versions Report

(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)

```
         Salt: 2015.5.2
         Python: 2.6.8 (unknown, Nov  7 2012, 14:47:45)
         Jinja2: unknown
       M2Crypto: 0.21.1
 msgpack-python: 0.1.12
   msgpack-pure: Not Installed
       pycrypto: 2.3
        libnacl: Not Installed
         PyYAML: 3.08
          ioflo: Not Installed
          PyZMQ: 2.1.9
           RAET: Not Installed
            ZMQ: 2.1.9
           Mako: Not Installed

```

We **use salt-ssh** to execute commands.

### LINKED ISSUE #34171: service.running state does not respect package requirement
### Description of Issue/Question

The following state can't be run in the test mode:

```
dbus:
  pkg:
    - installed
  service.running:
    - require:
        - pkg: dbus
```

It fails like this:

```
[ERROR   ] The named service dbus is not available
local:
----------
          ID: dbus
    Function: pkg.installed
      Result: None
     Comment: The following packages would be installed/updated: dbus
     Started: 13:53:27.894208
    Duration: 4293.316 ms
     Changes:
----------
          ID: dbus
    Function: service.running
      Result: False
     Comment: The named service dbus is not available
     Started: 13:53:32.188325
    Duration: 108.61 ms
     Changes:

Summary for local
------------
Succeeded: 1 (unchanged=1)
Failed:    1
------------
Total states run:     2
```

Unsurprisingly, it works as expected when not in dry run mode.

I made a docker image to replicate this easily:

```
docker build -t salt --build-arg SALT_VERSION=2016.3 https://gist.githubusercontent.com/bobrik/a7972c4cc9d8232f6b1ba9a0dff3a2cc/raw/e05680118cc9cd1703e27faf2017a8291d599493/Dockerfile && docker run --rm -it salt salt-call --local state.sls test test=true
```

The issue here is that `service.running` checks for available services when the required package is not yet installed, which is the case for all packages that have service file in them.

If there is a way to test this nicely, it should be documented. If there is no way to do this (which is sad), it should also be noted in `service` state docs.

* https://docs.saltstack.com/en/latest/ref/states/all/salt.states.service.html

### Versions Report

```
Salt Version:
           Salt: 2016.3.1

Dependency Versions:
           cffi: 0.8.6
       cherrypy: Not Installed
       dateutil: 2.2
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.3
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: 2.10
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: debian 8.4
        machine: x86_64
        release: 4.1.19-boot2docker
         system: Linux
        version: debian 8.4
```

### PR DESCRIPTION
**Pull request by prototux**
### What does this PR do?
It fixes the service.running call when using test=true and when the service isn't available (the common case is a new satefile that installs and runs a service being tested for the first time).

### What issues does this PR fix or reference?
Fixes #34171 and partialy fixes #34019

### Previous Behavior
When using test=true, it will return a failed state if the service is not available

### New Behavior
When using test=true, il will return none and a message if the service is not available

### Tests written?
No

### CODE REVIEW

File: salt/states/service.py
Comment (lorengordon) [inline salt/states/service.py]: Message should probably be, `Service {0} not present; if created in this state run, it would have been started', or something like that.

**Comment by lorengordon** [comment]
Might also want to do something similar for `service.dead`. Haven't looked at the code recently, but it used to have logic quite similar to `service.running`.

**Comment by prototux** [comment]
@lorengordon I've fixed the message as requested and added the same (with a slightly different message, obviously) for `service.dead`.

**Comment by lorengordon** [comment]
👍

**Comment by rallytime** [comment]
Hi @prototux Thanks for submitting this change. It looks like there are some test failures. Would you mind taking a look (don't worry about the pillar failures)? https://jenkins.saltstack.com/job/PR/job/salt-pr-rs-cent7-n/10814/

Also, @terminalmage would you mind looking this one over?

**Comment by prototux** [comment]
@rallytime it should be ok now (didn't tested that 'test' *was* in __opts__ before testing it's value), thanks for the feedback

File: salt/states/service.py
Comment (cachedout) [inline salt/states/service.py]: We could do `if __opts__.get('test'):` here

Comment (cachedout) [inline salt/states/service.py]: We could do `if __opts__.get('test'):` here

File: salt/states/service.py
Comment (cachedout) [inline salt/states/service.py]: Same here.

File: salt/states/service.py
Comment (prototux) [inline salt/states/service.py]: Ah, that seems better, thanks; there's probably the same thing to be said in #40696 i guess?

**Comment by cachedout** [comment]
We're just waiting on a review by @terminalmage before we get this in.

**Review by terminalmage — APPROVED** [review]

**Comment by terminalmage** [comment]
Looks good

**Comment by joniknsk** [comment]
Hi there!
In which version would this PR code released? Thanks!

**Comment by rallytime** [comment]
Hi @joniknsk - This is currently available in the `develop` branch, so it will be included in the next feature release. (Codenamed `Oxygen`.)

**Pull request merged**
